### PR TITLE
feat(auth): Implemented IdP config management APIs

### PR DESCRIPTION
**Pull request by hiranya911**
RELEASE NOTE: Added `OidcProviderConfig` and `SamlProviderConfig` classes for representing OIDC and SAML identity provider configurations.
RELEASE NOTE: Added new APIs to `FirebaseAuth` to facilitate creating, retrieving, updating and deleting identity provider configurations.

### CODE REVIEW

**Comment by hiranya911** [comment]
@egilmorez please review the public API ref docs in following classes:

* FirebaseAuth
* AuthProviderConfig
* OidcProviderConfig
* SamlProviderConfig
* OidcProviderConfigArgs
* SamlProviderConfigArgs
* ListProviderConfigsOptions

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 1]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
 
+        /// <summary>
+        /// Looks up an OIDC auth provider configuration by the provided ID.
+        /// </summary>
+        /// <returns>A task that completes with a <see cref="OidcProviderConfig"/>.</returns>
+        /// <exception cref="ArgumentException">If the provider ID is null, empty or does not
+        /// contain the <c>oidc.</c> prefix.</exception>
+        /// <exception cref="FirebaseAuthException">If the specified provider config does not
+        /// exist.</exception>
+        /// <param name="providerId">The provider ID corresponding to the OIDC provider config
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1175]: The ID of the OIDC provider config to return.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 2]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+
+        /// <summary>
+        /// Looks up an OIDC auth provider configuration by the provided ID.
+        /// </summary>
+        /// <returns>A task that completes with a <see cref="OidcProviderConfig"/>.</returns>
+        /// <exception cref="ArgumentException">If the provider ID is null, empty or does not
+        /// contain the <c>oidc.</c> prefix.</exception>
+        /// <exception cref="FirebaseAuthException">If the specified provider config does not
+        /// exist.</exception>
+        /// <param name="providerId">The provider ID corresponding to the OIDC provider config
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1191]: The ID of the OIDC provider config to return.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 3]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+
+        /// <summary>
+        /// Looks up a SAML auth provider configuration by the provided ID.
+        /// </summary>
+        /// <returns>A task that completes with a <see cref="SamlProviderConfig"/>.</returns>
+        /// <exception cref="ArgumentException">If the provider ID is null, empty or does not
+        /// contain the <c>saml.</c> prefix.</exception>
+        /// <exception cref="FirebaseAuthException">If the specified provider config does not
+        /// exist.</exception>
+        /// <param name="providerId">The provider ID corresponding to the SAML provider config
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1213]: The ID of the SAML provider config to return.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 4]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1229]: The ID of the SAML provider config to return.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 5]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+        /// <summary>
+        /// Creates a new auth provider configuration.
+        /// </summary>
+        /// <returns>A task that completes with an <see cref="AuthProviderConfig"/>.</returns>
+        /// <exception cref="ArgumentException">If <paramref name="args"/> is null or
+        /// invalid.</exception>
+        /// <exception cref="FirebaseAuthException">If an unexpected error occurs while creating
+        /// the provider configuration.</exception>
+        /// <param name="args">Arguments that describe the new provider configuration.</param>
+        /// <typeparam name="T">Type of <see cref="AuthProviderConfig"/> to be created.</typeparam>
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1252]: to create.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 6]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+        /// </summary>
+        /// <returns>A task that completes with an <see cref="AuthProviderConfig"/>.</returns>
+        /// <exception cref="ArgumentException">If <paramref name="args"/> is null or
+        /// invalid.</exception>
+        /// <exception cref="FirebaseAuthException">If an unexpected error occurs while creating
+        /// the provider configuration.</exception>
+        /// <param name="args">Arguments that describe the new provider configuration.</param>
+        /// <param name="cancellationToken">A cancellation token to monitor the asynchronous
+        /// operation.</param>
+        /// <typeparam name="T">Type of <see cref="AuthProviderConfig"/> to be created.</typeparam>
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1271]: to create.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 7]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+        /// Updates an existing auth provider configuration.
+        /// </summary>
+        /// <returns>A task that completes with the updated <see cref="AuthProviderConfig"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">If <paramref name="args"/> is null or
+        /// invalid.</exception>
+        /// <exception cref="FirebaseAuthException">If the specified provider config does not
+        /// exist or if an unexpected error occurs while performing the update.</exception>
+        /// <param name="args">Properties to be updated in the provider configuration.</param>
+        /// <typeparam name="T">Type of <see cref="AuthProviderConfig"/> to be updated.</typeparam>
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1292]: to update.

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs [hunk 8]
```diff
@@ -1161,6 +1164,236 @@ public async Task<FirebaseToken> VerifySessionCookieAsync(string sessionCookie)
...
+        /// <returns>A task that completes with the updated <see cref="AuthProviderConfig"/>.
+        /// </returns>
+        /// <exception cref="ArgumentException">If <paramref name="args"/> is null or
+        /// invalid.</exception>
+        /// <exception cref="FirebaseAuthException">If the specified provider config does not
+        /// exist or if an unexpected error occurs while performing the update.</exception>
+        /// <param name="args">Properties to be updated in the provider configuration.</param>
+        /// <param name="cancellationToken">A cancellation token to monitor the asynchronous
+        /// operation.</param>
+        /// <typeparam name="T">Type of <see cref="AuthProviderConfig"/> to be updated.</typeparam>
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs:1312]: to update.

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/AuthProviderConfig.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        /// <summary>
+        /// Gets the provider ID defined by the developer. For an OIDC provider, this is always
+        /// prefixed by <c>oidc.</c>. For a SAML provider, this is always prefixed by <c>saml.</c>.
+        /// </summary>
+        public string ProviderId { get; }
+
+        /// <summary>
+        /// Gets the The user-friendly display name of the configuration. This name is also used
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/AuthProviderConfig.cs:44]: typo: the The

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        /// <list type="bullet">
+        /// <item>
+        /// <description>To verify the provided issuer.</description>
+        /// </item>
+        /// <item>
+        /// <description>To determine the authentication/authorization endpoint during the OAuth
+        /// <c>id_token</c> authentication flow.</description>
+        /// </item>
+        /// <item>
+        /// <description>To retrieve the public signing keys via <c>jwks_uri</c> to verify the
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs:50]: "...from jwks_uri, which is used to verify the signature of the ID token."

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        /// <list type="bullet">
+        /// <item>
+        /// <description>To verify the provided issuer.</description>
+        /// </item>
+        /// <item>
+        /// <description>To determine the authentication/authorization endpoint during the OAuth
+        /// <c>id_token</c> authentication flow.</description>
+        /// </item>
+        /// <item>
+        /// <description>To retrieve the public signing keys via <c>jwks_uri</c> to verify the
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs:43]: "...from jwks_uri, which is used to verify the signature of the ID token."

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        /// OIDC provider's ID token signature.</description>
+        /// </item>
+        /// <item>
+        /// <description>To determine the <c>claims_supported</c> to construct the user attributes
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs:54]: To determine the <c>claims_supported</c>, which are passed through in the additional user info response.

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        /// OIDC provider's ID token signature.</description>
+        /// </item>
+        /// <item>
+        /// <description>To determine the <c>claims_supported</c> to construct the user attributes
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs:47]: To determine the <c>claims_supported</c>, which are passed through in the additional user info response.

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+                {
+                    mask.Add(entry.Key);
+                }
+            }
+
+            return mask;
+        }
+
+        /// <summary>
+        /// A class for making batch get requests to list a specific type of auth provider
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs:178]: GET

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+        /// configurations. An instance of this class is used by the Google API client to provide
```
Comment (kevinthecheung) [inline FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs:179]: configuration

Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs (hunk 14 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/ProviderConfigClient.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs (hunk 13 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs (hunk 12 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs (hunk 11 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfigArgs.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs (hunk 10 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/OidcProviderConfig.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/Providers/AuthProviderConfig.cs (hunk 9 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/Providers/AuthProviderConfig.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 8 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 7 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 6 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 5 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 4 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 3 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 2 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

File: FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs (hunk 1 above)
Comment (hiranya911) [reply FirebaseAdmin/FirebaseAdmin/Auth/FirebaseAuth.cs]: Done

**Pull request merged**